### DISCUSSION TITLE
How to install a Feature as Non-Root User in devcontainer.json

### DISCUSSION THREAD

**Discussion by andar1an** [devcontainers/community#13] [Q&A]
Is it possible to have a feature explicitly installed as the defined `remoteUser` in devcontainer.json? Could this be an issue if it is not working as intended?

I have tried to use the rust feature in the past, however, it causes spurious network errors when trying to use cargo because cargo is installed as root, but I am using vscode user. 

e.g. Rust does not install as vscode user here

```
"features": {
"ghcr.io/devcontainers/features/rust:1": {
	"version": "latest",
	"profile": "default"
},
"remoteUser": "vscode"
```

**Comment by Chuxel** [comment]
Are they errors network errors or file permissions?  There could be a bug in the Feature. It's supposed to be adding group privs to allow the non-root user to access it.

Also, what OS are you on?

**Comment by andar1an** [comment]
It reads as network errors, but it is permission issues. For example if you Run `Cargo install somecrate` you will hit the spurious network issue. The OS is not the issue for this. I have had this happen on many OS's and it is always the installation of Rust that was the problem. Installing Rust or using rust official image as base instead of using this feature is what I have been doing. I have even checked ownership of related rust folders in built containers

**Reply by andar1an** [reply]
This may be the problem: **https://github.com/devcontainers/features/blob/main/src/rust/install.sh#L227**. Is it setting permissions to a specific group, and doesn't account for a different user?

**Reply by andar1an** [reply]
Maybe you need to do something like, unless user is already added to rustlang group: 
```
if [ "${USERNAME}" = "auto" ] || [ "${USERNAME}" = "automatic" ]; then
    USERNAME=""
    POSSIBLE_USERS=("vscode" "node" "codespace" "$(awk -v val=1000 -F ":" '$3==val{print $1}' /etc/passwd)")
    for CURRENT_USER in "${POSSIBLE_USERS[@]}"; do
        if id -u ${CURRENT_USER} > /dev/null 2>&1; then
            chmod -R u+r+w "${RUSTUP_HOME}" "${CARGO_HOME}"
            break
        fi
    done
    if [ "${USERNAME}" = "" ]; then
        chmod -R g+r+w "${RUSTUP_HOME}" "${CARGO_HOME}"
    fi
elif [ "${USERNAME}" = "none" ] || ! id -u ${USERNAME} > /dev/null 2>&1; then
    chmod -R g+r+w "${RUSTUP_HOME}" "${CARGO_HOME}"
fi
```

**Reply by andar1an** [reply]
In hindsight can probably just make sure user is in rustlang group.

**Reply by Chuxel** [reply]
Yeah, that's what the script does - so not sure what might be causing priv issues.  

https://github.com/devcontainers/features/blob/main/src/rust/install.sh#L175

Wonder if the rust install is ignoring `umask` now so the privs aren't set to group read/write. That's intended to speed things up by avoiding a chmod post install, but if privs are restored from an untar, that can stop working.
